src/Interface/ComputationalRelation.agda

```diff
@@ -5,68 +5,130 @@ module Interface.ComputationalRelation where
 open import Prelude
 open import Data.Maybe.Properties using (map-nothing)
 open import Interface.STS public
+open import Effect.Functor
+open import Effect.Applicative
+open import Effect.Monad
 
 private variable
   C S Sig : Set
+  Err Err₁ Err₂ : Set
   c : C
   s s' s'' : S
   sig : Sig
 
+data ComputationResult {a : Level} (Err : Set) (R : Set a) : Set a where
+  success : R → ComputationResult Err R
+  failure : Err → ComputationResult Err R
+
+isFailure : ∀ {a} {E} {A : Set a} → ComputationResult E A → Set a
+isFailure x = ∃[ e ] x ≡ failure e
+
+instance
+  Bifunctor-ComputationResult : ∀ {a : Level} → Bifunctor {_} {a} ComputationResult
+  Bifunctor-ComputationResult .bimap _ f (success x) = success $ f x
+  Bifunctor-ComputationResult .bimap f _ (failure x) = failure $ f x
+  
+  Functor-ComputationResult : ∀ {E : Set} → Functor (ComputationResult E)
+  Functor-ComputationResult ._<$>_ f (success x) = success $ f x
+  Functor-ComputationResult ._<$>_ _ (failure x) = failure x
+
+  Applicative-ComputationResult : ∀ {E : Set} → Applicative (ComputationResult E)
+  Applicative-ComputationResult .pure = success
+  Applicative-ComputationResult ._<*>_ (success f) x = f <$> x
+  Applicative-ComputationResult ._<*>_ (failure e) _ = failure e
+
+  Monad-ComputationResult : ∀ {E : Set} → Monad (ComputationResult E)
+  Monad-ComputationResult .return = success
+  Monad-ComputationResult ._>>=_ (success a) m = m a
+  Monad-ComputationResult ._>>=_ (failure e) _ = failure e
+
 module _ (STS : C → S → Sig → S → Set) where
 
-  ExtendedRel : C → S → Sig → Maybe S → Set
-  ExtendedRel c s sig (just s') = STS c s sig s'
-  ExtendedRel c s sig nothing   = ∀ s' → ¬ STS c s sig s'
+  ExtendedRel : C → S → Sig → ComputationResult Err S → Set
+  ExtendedRel c s sig (success s') = STS c s sig s'
+  ExtendedRel c s sig (failure _ ) = ∀ s' → ¬ STS c s sig s'
+
+  map-failure : ∀ {A B C : Set} {f : A → B} {x : C} {ma} → ma ≡ failure x → map f ma ≡ failure x
+  map-failure refl = refl
+
+  success-maybe : ∀ {Err R : Set} → ComputationResult Err R → Maybe R
+  success-maybe (success x) = just x
+  success-maybe (failure _) = nothing
+
+  failure-maybe : ∀ {Err R : Set} → ComputationResult Err R → Maybe Err
+  failure-maybe (success _) = nothing
+  failure-maybe (failure x) = just x
```

Comment:
Those just relate to `ComputationResult`, so they should not be in this `module` block